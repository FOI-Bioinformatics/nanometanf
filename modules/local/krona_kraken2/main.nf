process KRONA_KRAKEN2 {
    tag "$meta.id"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/krona:2.8.1--pl5321hdfd78af_1' :
        'biocontainers/krona:2.8.1--pl5321hdfd78af_1' }"

    input:
    tuple val(meta), path(kraken_report)

    output:
    tuple val(meta), path("*.html"), emit: html
    path "versions.yml"            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    # Convert Kraken2 report to Krona format
    # Kraken2 format: %reads, #reads_clade, #reads_taxon, rank, taxid, name
    # Krona format: count taxid [taxid taxid...]

    # Import Krona taxonomy database
    ktUpdateTaxonomy.sh

    # Convert Kraken2 report to Krona input
    awk -F'\\t' '\$4 != "U" {print \$3"\\t"\$5}' ${kraken_report} > ${prefix}.krona.txt

    # Generate interactive Krona plot
    ktImportTaxonomy \\
        -o ${prefix}.krona.html \\
        -t 5 \\
        -m 3 \\
        ${prefix}.krona.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        krona: \$( ktImportTaxonomy 2>&1 | sed -n 's/^.*KronaTools //p' | sed 's/ - .*\$//' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.krona.html

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        krona: 2.8.1
    END_VERSIONS
    """
}
