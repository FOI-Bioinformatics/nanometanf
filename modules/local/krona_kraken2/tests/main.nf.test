nextflow_process {

    name "Test Process KRONA_KRAKEN2"
    script "../main.nf"
    process "KRONA_KRAKEN2"

    tag "modules"
    tag "modules_local"
    tag "krona_kraken2"

    test("krona_kraken2 - stub run") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [id: 'test_sample'],
                    file(params.test_data['generic']['txt']['kraken2_report'], checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
