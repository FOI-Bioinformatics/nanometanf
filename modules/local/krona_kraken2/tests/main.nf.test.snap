{
    "krona_kraken2 - stub run": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.krona.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,9969c8ca126b077b73bd6a7250fa2ef2"
                ],
                "html": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.krona.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9969c8ca126b077b73bd6a7250fa2ef2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:26:19.597094"
    }
}